packages/@lwc/engine-server/src/renderer.ts

```diff
@@ -330,23 +338,114 @@ function isConnected(node: HostNode) {
+    if (node.type !== 'element') {
+        return;
+    }
+
+    let savedCallback = callback;
+
+    if (useCaptureOrOptions) {
+        if (useCaptureOrOptions === true || useCaptureOrOptions.capture) {
+            // Capturing listeners aren't meaningful in SSR, since the capture phase
+            // is not supported. However, we don't need to break apps that register
+            // these listeners.
+            return;
+        }
+        if (useCaptureOrOptions.once) {
+            savedCallback = (...args) => {
+                try {
+                    callback(...args);
+                } finally {
+                    removeEventListener(node, type, savedCallback);
+                }
+            };
+        }
+
+        // In the SSR context, `passive` event listeners are not meaningful and are ignored.
+        // Additionally, we do not currently support AbortSignals, so the `signal` option
+        // is also ignored.
+    }
 
-// Noop on SSR.
-const removeEventListener = noop as (
-    target: HostNode,
+    if (!(type in node.eventListeners)) {
+        node.eventListeners[type] = new Set();
+    }
+    node.eventListeners[type].add(savedCallback);
+}
+
+function removeEventListener(
+    node: N,
     type: string,
-    callback: EventListener,
-    options?: AddEventListenerOptions | boolean
-) => void;
+    callback: EventListener
+    // captured listeners aren't supported in SSR, so options are ignored
+): void {
+    if (node.type !== 'element') {
+        return;
+    }
+    const eventListeners = node.eventListeners[type];
+    if (eventListeners) {
+        eventListeners.delete(callback);
+    }
+}
+
+type EventProperty = keyof Event;
+function dispatchEvent(target: HostNode, event: Event): boolean {
+    if (target.type !== 'element') {
+        return true;
+    }
+
+    let currentNode: HostElement | HostShadowRoot | null = target;
+    let stop = false;
+    let stopImmediately = false;
+
+    const stopPropagation = () => {
+        stop = true;
+    };
+    const stopImmediatePropagation = () => {
+        stop = true;
+        stopImmediately = true;
+    };
+    const eventProxy = new Proxy<Event>(event, {
+        get(eventObj, property: EventProperty) {
+            if (property === 'stopPropagation') {
+                return stopPropagation;
+            } else if (property === 'stopImmediatePropagation') {
+                return stopImmediatePropagation;
+            }
+            return eventObj[property];
+        },
+    });
+
+    do {
+        if (currentNode.type === HostNodeType.Element) {
+            const callbacks: Set<EventListener> | undefined =
+                currentNode.eventListeners[event.type];
+            if (callbacks) {
+                for (const callback of callbacks) {
+                    if (!stopImmediately) {
+                        callback(eventProxy);
+                    }
+                }
+            }
+        }
+        currentNode = currentNode.parent;
```

Comment:
The event should only invoke all the ancestor elements if `bubbles` is set to `true`.

Reply:
The behavior of `bubbles: false, composed: true` is especially counter-intuitive here. I had to look at [PM's post](https://pm.dartus.fr/blog/a-complete-guide-on-shadow-dom-and-event-propagation/#escaping-the-shadow-trees-using-composed) a few times before I got it. 😆 

> [W]hen the dispatched event is composed only, the event propagates from one host element to another [...] without propagating through the intermediary nodes.

Reply:
I unresolved this convo, because I don't think the logic is correct for the case @nolanlawson mentioned `bubbles: false, composed: true`. If `bubbles` is `false`, `stop` is `true` which prevents the event to propagate to the closest shadow root ancestor.

Reply:
Thank you for this, btw. I'll make sure there's good test coverage here.